### ISSUE TITLE
Consul cert storage format

### ISSUE THREAD

**Issue by cpredmann** [eBay/fabio#171]
What is the format that fabio is expecting for cert storage in Consul?
I've been trying a few different things, but keep seeing `Store has certificates for [""]`, and it clearly isn't reading them.  Trying the cert/key using the file storage works as expected, so the certs are valid.
As for path, I've been using v1/kv/ssl/foo.com/cert = ---BEGIN CERTIFICATE--- and v1/kv/ssl/foo.com/key = ----BEGIN RSA PRIVATE KEY---

**Comment by magiconair** [comment]
PEM but your path is incorrect. Use either `v1/kv/ssl/foo.com.pem` if you have key and cert in one file or `v1/kv/ssl/foo.com-cert.pem` and `v1/kv/ssl/foo.com-key.pem` if they are separate.

Let me know whether that helps.

From `fabio.properties`:
```
# Path
#
...
#
# TLS certificates are stored either in one or two files:
#
#   www.example.com.pem or www.example.com-{cert,key}.pem
#
# TLS certificates are loaded in alphabetical order and the first certificate
# is the default for clients which do not support SNI.
#
...
# Consul
#
...
# The filenames follow the same rules as for the path source.
...
```

**Comment by cpredmann** [comment]
Thanks for the quick response, and that makes sense, although I still must be missing something.
In the logs, I see: `[181840.928314] fabio[5]:     "CertSources": {
[181840.928541] fabio[5]:         "cert": {
[181840.928814] fabio[5]:             "Name": "cert",
[181840.929040] fabio[5]:             "Type": "consul",
[181840.929301] fabio[5]:             "CertPath": "http://my.consul.server:8500/v1/kv/ssl/*.foo.bar.com-cert.pem?token=my-consul-token",
[181840.929543] fabio[5]:             "KeyPath": "http://my.consul.server:8500/v1/kv/ssl/*.foo.bar.com-key.pem?token=my-consul-token",
[181840.929886] fabio[5]:             "ClientCAPath": "",
[181840.930148] fabio[5]:             "CAUpgradeCN": "",
[181840.930412] fabio[5]:             "Refresh": 3000000000,
[181840.930675] fabio[5]:             "Header": null
[181840.930914] fabio[5]:         }
[181840.931165] fabio[5]:     },`
Where those key and cert paths are reachable using the specified token.  They are wildcard certs, and I've tried with and without *. to see if that makes a difference, I'm sure I'm just overlooking something simple

**Comment by magiconair** [comment]
You need to provide only the base path in the properties. Then store the certificates under these filenames and fabio will pick them up. So you'd want something like this in your `fabio.properties`

    cs=cert;type=consul;cert=http://my.consul.server:8500/v1/kv/ssl&token=my-consul-token

**Comment by cpredmann** [comment]
Great, thanks again

**Comment by cpredmann** [comment]
After getting it to work without a token, but then failing with a supplied token, it looks like there may be an issue with the Consul path when using a token.  In consul_source.go parseConsulURL, the token is parsed as a query parameter from the config, but still remains as part of the URL.  So, when using `key = u.RequestURI()[len(stripPrefix):]`, the Consul request is `v1/key/ssl?token=my-token?consistent=&recurse=&token=<hidden>` where token is added twice (with incorrect formatting).  Changing to `key = u.Path[len(stripPrefix):]` looks to resolve correctly, though

**Comment by magiconair** [comment]
You found a bug? I'll have a look later. Thx

**Comment by magiconair** [comment]
I've filed this as a separate bug: #172 Fix is ready. Thx for finding this.

**Comment by magiconair** [comment]
I've pushed the fix for #172 to master. Could you please check whether that fixes your issue? I'm going to close this question since we're now talking about a bug and I prefer not to morph discussions.
